lib/charms/kubernetes_charm_libraries/v0/multus.py

```diff
@@ -0,0 +1,413 @@
+                    "NetworkAttachmentDefinition resource not found. "
+                    "You may need to install Multus CNI."
+                )
+            else:
+                raise KubernetesMultusError(
+                    f"Unexpected outcome when retrieving network attachment definition {name}"
+                )
+
+    def create_network_attachment_definition(
+        self, network_attachment_definition: GenericNamespacedResource
+    ) -> None:
+        """Creates a NetworkAttachmentDefinition.
+
+        Args:
+            network_attachment_definition: NetworkAttachmentDefinition object
+        """
+        try:
+            self.client.create(obj=network_attachment_definition, namespace=self.namespace)  # type: ignore[call-overload]  # noqa: E501
+        except ApiError:
+            raise KubernetesMultusError(
+                f"Could not create NetworkAttachmentDefinition "
+                f"{network_attachment_definition.metadata.name}"  # type: ignore[union-attr]
+            )
+        logger.info(
+            f"NetworkAttachmentDefinition {network_attachment_definition.metadata.name} created"  # type: ignore[union-attr]  # noqa: E501, W505
+        )
+
+    def delete_network_attachment_definition(self, name: str) -> None:
+        """Deletes network attachment definition based on name.
+
+        Args:
+            name: NetworkAttachmentDefinition name
+        """
+        try:
+            self.client.delete(
+                res=NetworkAttachmentDefinition, name=name, namespace=self.namespace
+            )
+        except ApiError:
+            raise KubernetesMultusError(f"Could not delete NetworkAttachmentDefinition {name}")
+        logger.info(f"NetworkAttachmentDefinition {name} deleted")
+
+    def patch_statefulset(
+        self,
+        name: str,
+        network_annotations: list[NetworkAnnotation],
+        containers_requiring_net_admin_capability: list[str],
+    ) -> None:
+        """Patches a statefulset with multus annotation.
+
+        Args:
+            name: Statefulset name
+            network_annotations: list of network annotations
+            containers_requiring_net_admin_capability: Containers requiring NET_ADMIN capability
+        """
+        if not network_annotations:
+            logger.info("No network annotations were provided")
+            return
+        try:
+            statefulset = self.client.get(res=StatefulSet, name=name, namespace=self.namespace)
+        except ApiError:
+            raise KubernetesMultusError(f"Could not get statefulset {name}")
+        statefulset.spec.template.metadata.annotations["k8s.v1.cni.cncf.io/networks"] = json.dumps(  # type: ignore[attr-defined]  # noqa: E501
+            [network_annotation.dict() for network_annotation in network_annotations]
+        )
+        for container_name in containers_requiring_net_admin_capability:
+            container_index = self.get_container_index_from_name(
```

Comment:
shouldn't we handle possible exceptions? or are we sure the statefulset + containers will be found?

Reply:
There is exception handling around the `get` inside of the `get_container_index_from_name`. So if the statefulset is not found, a custom exception `KubernetesMultusError` would be raised.